### ISSUE TITLE
r_binary_packages and r_github_packages sections not being processed

### ISSUE THREAD

**Issue by niranjv** [travis-ci/travis-ci#3389]
My .travis.yml file is at:
https://github.com/niranjv/ucscamsms2015/blob/develop/.travis.yml

It is based on the 2 .travis.yml files below:
* https://github.com/hadley/devtools/blob/master/.travis.yml
* https://github.com/hadley/r-pkgs/blob/master/.travis.yml

My R build script is failing because the required R packages are missing. These packages are specified in the `r_binary_packages` and `r_github_packages` sections in my .travis.yml file but these sections are not being processed at all. There is no mention in the Travis log files of any attempt to even install the R packages (https://travis-ci.org/niranjv/ucscamsms2015/builds/54648610). It's as if these sections didn't exist my .travis.yml file.

I tried a few things like verifying that the .travis.yml file is valid YAML, explicitly specifying the CRAN mirror, changing the order of statements, etc. but nothing made a difference. There is no reference to this issue in either the Travis help docs or in Stack Overflow and I am out of ideas on how to fix this issue. Can someone help? Thanks!

cc: @craigcitro, @eddelbuettel, @hadley

**Comment by craigcitro** [comment]
hi @niranjv,

ah, i know what's going wrong here. we install all the dependencies as part of the `install` step, which you've overridden here (with the `sudo apt-get install ...` line). three things:

1/ you can add an `apt_packages` list, which should install that for you. [example](https://github.com/craigcitro/r-travis/blob/new_config/.travis.yml#L15-L17)
2/ there are `before_install` and `after_install` sections you can use if you need some installation that we can't handle.
3/ i wonder if we should add something to our `script:` section that notifies the user if the `install:` section hasn't run?

one more question: you mention you couldn't find help for this; just to confirm, you found the [R page in the travis docs](http://docs.travis-ci.com/user/languages/r/), but that didn't answer your question? (i just want to make sure that page is sufficiently discoverable.)

**Comment by craigcitro** [comment]
PS @niranjv if you can try moving that line and report back, that would be awesome.

**Comment by niranjv** [comment]
Thanks for the info.

* I did find the R page in the Travis docs, but I did not see any warning about R packages not being installed because the `install` section was overridden
* Moving the apt-get line from the `install` section to the `apt_packages` section in my .travis.yml file fixed this issue and both xelatex and all R packages specified in the .travis.yml file seem to be installed correctly...
* ... but Travis now fails with the error below that not all dependencies were installed:
```
The command "Rscript -e 'options(repos = "http://cran.rstudio.com"); tryCatch({   deps <- devtools::install_deps(dependencies = TRUE) }, error = function(e) {   message(e);   q(status=1) }); if (!all(deps %in% installed.packages())) {  q(status = 1, save = "no") }'" failed and exited with 1 during .
```

Full log is at https://travis-ci.org/niranjv/ucscamsms2015/builds/54658675 

If it helps, I'm not trying to build an R package here. Just trying to use knitr and xelatex to convert some *.rmd files to *.tex files and then to a PDF file.

**Comment by craigcitro** [comment]
ah -- that makes sense. in this case, i think you *do* want to override the `script` section, but *not* the `install` section. :smile: 

the `script` section by default builds and tests your package; in your case, you should just drop in whatever commands you use to do the `rmd` -> `tex` -> `pdf` conversion.

i think you're the first person doing this on top of the travis support, so once you do, i'm going to turn your `.travis.yml` into a sample in the docs.

**Comment by niranjv** [comment]
All packages in the `apt_packages`, `r_binary_packages` and `r_github_packages` sections are installed correctly. But nothing in the `script` section is being run. Either Travis is ignoring this section or it is failing before getting to this section. Travis fails while checking if all dependencies of my "package" are installed correctly. I don't know why this check is being run. No info in given on what the dependencies are or which packages are missing, so I'm stuck again.

Error msg in the log:

The command "Rscript -e 'options(repos = "http://cran.rstudio.com"); tryCatch({   deps <- devtools::install_deps(dependencies = TRUE) }, error = function(e) {   message(e);   q(status=1) }); if (!all(deps %in% installed.packages())) {  q(status = 1, save = "no") }'" failed and exited with 1 during .

Your build has been stopped.

**Comment by BanzaiMan** [comment]
@niranjv The build stops because the failure occurs inside the `install` whose command has the `assert: true` (default for the `install` section).

This is working as designed.

**Comment by craigcitro** [comment]
@BanzaiMan but why is it running the built-in `script` section if he's provided his own to override it?

@niranjv i think i'll add an option for "this is not an R package", to make cases like this one a little more explicit. i'm hoping to get a chance to knock out a few travis-related fixes soon (sick kids == personal projects grind to a halt :smile: )

**Comment by BanzaiMan** [comment]
What built-in `script`? R's built-in `script` starts with this announcement (https://github.com/travis-ci/travis-build/blob/edaedf7a024b610f0ce67af548961ac0209eb4f9/lib/travis/build/script/r.rb#L133): `Building with: R CMD build ${R_BUILD_ARGS}`.

I don't see that in https://travis-ci.org/niranjv/ucscamsms2015/builds/54831439. The build doesn't get to the `script` section at all.

The failure comes from here: https://github.com/travis-ci/travis-build/blob/edaedf7a024b610f0ce67af548961ac0209eb4f9/lib/travis/build/script/r.rb#L290

If a failure in this command can be safely ignored, you should have `assert: false`.

**Comment by craigcitro** [comment]
ah, @banzaiman, you're right - I'd forgotten install_deps was in install. it should fail the build if that fails. I just need to add an option to skip it altogether in cases like this.

@niranjv I think you're stuck until I get a chance to tweak a bit, sadly. can you stick with r-Travis and I'll ping this thread when I get a fix in?

**Comment by eddelbuettel** [comment]
I agree.  @niranjv has a use case entirely outside of what this is trying to do here.  

By all means take inspiration from our script.  But if they fail for _a use case clearly outside the spec they were designed for_ we are not all that responsible for the failure.

**Comment by niranjv** [comment]
OK, thanks for the clarification. I will proceed with r-travis.

**Comment by jpritikin** [comment]
I'm running into a similar problem. I need my own custom install hook because my package has a `DESCRIPTION.in` instead of `DESCRIPTION`. I used `apt_packages`. Why did my job [fail](https://travis-ci.org/OpenMx/OpenMx/jobs/198849770)? Thank you.

**Comment by jpritikin** [comment]
I also tried using lower-level apt-get commands, but that [failed](https://travis-ci.org/OpenMx/OpenMx/jobs/198879375) too. How do I debug this?

CC: @craigcitro, @hadley, and @jimhester

**Comment by jimhester** [comment]
@jprintkin You needed to use `r_binary_packages` to install R binary packages. If you have overridden the `install:` step and are trying to do this manually all packages in the marutter PPA are _lowercase_ package names prefixed with `r-cran-`, e.g. `r-cran-rcpp`.

**Comment by jpritikin** [comment]
@jimhester Ah! I didn't realize that the package names were prefixed, etc. Thank you very much for pointing this out. Things are working now.
